File: Scala/src/main/scala/TextbookTrack/Chapter02/BA2H.scala
package TextbookTrack.Chapter02

object BA2H {
  private def calcHammingDistance(s1: String, s2: String): Int = s1.lazyZip(s2).count{ case (c1, c2) => c1 != c2 }

  private def distanceBetweenPatternAndText(text: String, pattern: String, k: Int): Int =
    text.sliding(k).foldLeft(k)((acc, kMer) => math.min(acc, calcHammingDistance(pattern, kMer)))

  def distanceBetweenPatternAndCollectionOfTexts(texts: List[String], pattern: String, k: Int): Int =
    texts.foldLeft(0)((acc, text) => acc + distanceBetweenPatternAndText(text, pattern, k))

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val pattern: String = reader.next()
    val k: Int = pattern.length
    val texts: List[String] = reader.next().split(" ").toList
    val result: Int = distanceBetweenPatternAndCollectionOfTexts(texts, pattern, k)
    println(result)
  }
}

File: OCaml/src/textbook_track/chapter02/ba2h.ml
let calc_hamming_distance (s1 : string) (s2 : string) : int =
    Seq.fold_left2 (fun acc c1 c2 -> if c1 <> c2 then acc + 1 else acc) 0 (String.to_seq s1) (String.to_seq s2)


let calc_distance_between_pattern_and_text (text : string) (pattern : string) (k : int) : int =
    let n = String.length text in
    let substrings = Seq.init (n - k + 1) (fun ix -> String.sub text ix k) in
    Seq.fold_left (fun acc k_mer -> min acc (calc_hamming_distance k_mer pattern)) k substrings


let distance_between_patterns_and_texts (texts : string list) (pattern : string) (k : int) : int =
    List.fold_left (fun acc text -> acc + calc_distance_between_pattern_and_text text pattern k) 0 texts


let () =
    let pattern = read_line () in
    let k = String.length pattern in
    let texts = String.split_on_char ' ' @@ read_line () in
    let result = distance_between_patterns_and_texts texts pattern k in
    print_int result; print_newline ()
